File: backend/app/schemas/audit_room.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class AuditRoomFolderResponse(BaseModel):
    id: str
    label: str
    control_ref_prefix: str
    item_count: int


class AuditRoomEvidenceItem(BaseModel):
    id: str
    name: str
    type: str
    control_id: str
    clause_ref: str
    organization_id: str
    question_id: Optional[str] = None
    answer_id: Optional[str] = None
    validity_days: Optional[int] = None
    freshness_status: str
    created_at: Optional[datetime] = None


class AuditRoomSearchRequest(BaseModel):
    query: str = Field(min_length=1)


class AuditRoomSearchResult(BaseModel):
    id: str
    clause_ref: Optional[str] = None
    title: Optional[str] = None
    content: Optional[str] = None
    relevance_score: float


class AuditRoomSearchResponse(BaseModel):
    query: str
    results: List[AuditRoomSearchResult]


class AuditRoomBinderRequest(BaseModel):
    title: str = Field(min_length=3)
    description: Optional[str] = None
    selected_evidence_ids: List[str] = Field(min_items=1)


class AuditRoomBinderResponse(BaseModel):
    job_id: str
    status_url: str


class AuditRoomBinderStatusResponse(BaseModel):
    job_id: str
    status: str
    progress: int
    message: Optional[str] = None
    download_url: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

File: frontend/src/api/auditRoom.ts
import client from './client'

export interface AuditRoomFolderResponse {
  id: string
  label: string
  control_ref_prefix: string
  item_count: number
}

export interface AuditRoomEvidenceItem {
  id: string
  name: string
  type: string
  control_id: string
  clause_ref: string
  organization_id: string
  question_id?: string | null
  answer_id?: string | null
  validity_days?: number | null
  freshness_status: string
  created_at?: string | null
}

export interface AuditRoomSearchResult {
  id: string
  clause_ref?: string | null
  title?: string | null
  content?: string | null
  relevance_score: number
}

export interface AuditRoomSearchResponse {
  query: string
  results: AuditRoomSearchResult[]
}

export interface AuditRoomBinderRequest {
  title: string
  description?: string
  selected_evidence_ids: string[]
}

export interface AuditRoomBinderResponse {
  job_id: string
  status_url: string
}

export interface AuditRoomBinderStatusResponse {
  job_id: string
  status: string
  progress: number
  message?: string
  download_url?: string
  created_at?: string
  updated_at?: string
}

export const getAuditRoomFolders = async (): Promise<AuditRoomFolderResponse[]> => {
  const { data } = await client.get<AuditRoomFolderResponse[]>('/audit-room/folders')
  return data
}

export const getAuditRoomFolderEvidences = async (controlId: string): Promise<AuditRoomEvidenceItem[]> => {
  const { data } = await client.get<AuditRoomEvidenceItem[]>(`/audit-room/folders/${encodeURIComponent(controlId)}/evidences`)
  return data
}

export const searchAuditRoom = async (query: string): Promise<AuditRoomSearchResponse> => {
  const { data } = await client.post<AuditRoomSearchResponse>('/audit-room/search', { query })
  return data
}

export const createAuditRoomBinder = async (
  payload: AuditRoomBinderRequest,
): Promise<AuditRoomBinderResponse> => {
  const { data } = await client.post<AuditRoomBinderResponse>('/audit-room/binder', payload)
  return data
}

export const getAuditRoomBinderStatus = async (
  jobId: string,
): Promise<AuditRoomBinderStatusResponse> => {
  const { data } = await client.get<AuditRoomBinderStatusResponse>(`/audit-room/binder/${encodeURIComponent(jobId)}/status`)
  return data
}

export const downloadAuditRoomBinder = async (downloadUrl: string): Promise<Blob> => {
  const normalizedUrl = downloadUrl.startsWith('/api/v1') ? downloadUrl.replace(/^\/api\/v1/, '') : downloadUrl
  const response = await client.get<Blob>(normalizedUrl, { responseType: 'blob' })
  return response.data
}
